src/shims/sync.rs

```diff
@@ -0,0 +1,519 @@
+    ecx.write_scalar(writers.into(), writers_place.into())
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn pthread_mutexattr_init(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let default_kind = this.eval_libc("PTHREAD_MUTEX_DEFAULT")?;
+        mutexattr_set_kind(this, attr_op, default_kind)?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_settype(
+        &mut self,
+        attr_op: OpTy<'tcx, Tag>,
+        kind_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let kind = this.read_scalar(kind_op)?.not_undef()?;
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")?
+            || kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")?
+            || kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")?
+        {
+            mutexattr_set_kind(this, attr_op, kind)?;
+        } else {
+            let einval = this.eval_libc_i32("EINVAL")?;
+            return Ok(einval);
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_destroy(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        mutexattr_set_kind(this, attr_op, ScalarMaybeUndef::Undef)?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutex_init(
+        &mut self,
+        mutex_op: OpTy<'tcx, Tag>,
+        attr_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        let kind = if this.is_null(attr)? {
+            this.eval_libc("PTHREAD_MUTEX_DEFAULT")?
+        } else {
+            mutexattr_get_kind(this, attr_op)?.not_undef()?
+        };
+
+        mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
+        mutex_set_kind(this, mutex_op, kind)?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutex_lock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
```

Comment:
The docs do not say that passing a null ptr returns EINVAL, so conservatively I'd prefer to just make that UB (which should happen implicitly if you just remove this branch).
Is this behavior guaranteed in practice?

I found this in the docs:

> EINVAL
>    The value specified by mutex does not refer to an initialized mutex object. 

but surely, if I just pass `0xdeadbeef`, there is no guarantee that I will get EINVAL, that is just UB. Why is the null pointer special?

Reply:
(Same for all the other NULL ptr checks.)